pkg/envoy/registry/services.go

```diff
@@ -27,34 +33,145 @@ func (e ExplicitProxyServiceMapper) ListProxyServices(p *envoy.Proxy) ([]service
 
 // KubeProxyServiceMapper maps an Envoy instance to services in a Kubernetes cluster.
 type KubeProxyServiceMapper struct {
-	KubeController k8s.Controller
+	kubeController k8s.Controller
+	kubeEvents     chan interface{}
+	servicesForCN  map[certificate.CommonName][]service.MeshService
```

Comment:
I suspect that this may need to be a sync.Map or something similar since we may attempt to write to this in one goroutine and read from it in another.

Reply:
Good catch. I added synchronization around that and a test that passes with `-race` (that I confirmed failed without the locking).